File: solutions/3805-maximize-active-section-with-trade-i/maximize-active-section-with-trade-i.go
func maxActiveSectionsAfterTrade(s string) int {
	n := len(s)
	cnt1 := 0
	for _, c := range s {
		if c == '1' {
			cnt1++
		}
	}

	i := 0
	bestGain := 0
	prev := -1 << 31
	cur := 0

	for i < n {
		start := i
		for i < n && s[i] == s[start] {
			i++
		}
		if s[start] == '0' {
			cur = i - start
			if prev != -1<<31 {
				bestGain = max(bestGain, prev+cur)
			}
			prev = cur
		}
	}

	return cnt1 + bestGain
}

File: solutions/3805-maximize-active-section-with-trade-i/maximize-active-section-with-trade-i.rs
impl Solution {
    pub fn max_active_sections_after_trade(s: String) -> i32 {
        let n = s.len();
        let cnt1 = s.chars().filter(|&c| c == '1').count() as i32;

        let mut i = 0;
        let mut best_gain = 0;
        let mut prev = i32::MIN;
        let mut cur = 0;
        let bytes = s.as_bytes();

        while i < n {
            let start = i;
            while i < n && bytes[i] == bytes[start] {
                i += 1;
            }
            if bytes[start] == b'0' {
                cur = (i - start) as i32;
                if prev != i32::MIN {
                    best_gain = best_gain.max(prev + cur);
                }
                prev = cur;
            }
        }

        cnt1 + best_gain
    }
}